### PR TITLE
fix(cli): Markdown Link Sanitization in the InlineMarkdownRenderer

### PR DESCRIPTION
**Pull request by davideast**
## TLDR

This pull request significantly refactors the inline markdown renderer to be more robust, secure, and maintainable. It replaces a monolithic regex parser with a data-driven approach, improves link sanitization, and fixes a functional regression in italics parsing.

### Refactoring to a Data-Driven Approach

The previous implementation relied on a single, complex regular expression and a large `while` loop to parse and render markdown. This has been replaced with a `MARKDOWN_RULES` array that defines each markdown element as a separate entity with its own named-capture-group regex. This makes the parsing logic more modular, easier to read, and extensible. Each rule has a corresponding handler for rendering (in React) and stripping (for plain text length calculation), which cleans up the control flow significantly.

### Security and Sanitization

A major improvement is the introduction of proper link sanitization. The `extractLinkContent` function now explicitly checks for safe protocols (`https`, `http`, `ftp`, `mailto`) before rendering a link. This prevents unsafe protocols like `javascript:` from being rendered as active links, addressing a potential security vulnerability.

### Bug Fixes and Regressions

- **Italics Parsing:** This PR fixes a high-severity functional regression where the italics regex (`\*.+?\*|_.+?_`) was too greedy and would incorrectly format parts of words (e.g., `file_name` or `a*b*c`). The updated regex, `(?<italic>(?<![a-zA-Z0-9])(?:\*(?!\s).+?(?<!\s)\*|_(?!\s).+?(?<!\s)_)(?![a-zA-Z0-9]))`, uses negative lookarounds to ensure that emphasis is only applied to whole words, aligning with CommonMark specifications.

- **FTP Link Handling:** The previous implementation had inconsistent handling of `ftp` links. The new `isSafe` check in `extractLinkContent` now correctly includes `ftp` as a safe protocol, ensuring that `ftp` links are rendered consistently with other link types.

## Testing Matrix

|          | 🍏  | 🪟  | 🐧  |
| -------- | --- | --- | --- |
| npm run  | ❓  | ❓  | ✅   |
| npx      | ❓  | ❓  | ❓  |
| Docker   | ❓  | ❓  | ❓  |
| Podman   | ❓  | -   | -   |
| Seatbelt | ❓  | -   | -   |

### CODE REVIEW

File: packages/cli/src/ui/utils/__snapshots__/InlineMarkdownRenderer.test.tsx.snap [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+exports[`RenderInline > renders italic text correctly 1`] = `
+"This is
+italic
+."
+`;
+
+exports[`RenderInline > renders strikethrough text correctly 1`] = `
+"This is
+strikethrough
```
Comment (jacob314) [inline packages/cli/src/ui/utils/__snapshots__/InlineMarkdownRenderer.test.tsx.snap]: this screenshot is misleading as there is no strikethrough. same for others that don't really render. please remove those and test these edge cases via test cases rather than adding snapshots that look misleading.

File: packages/cli/src/ui/utils/InlineMarkdownRenderer.test.tsx [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+ * Copyright 2025 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { describe, it, expect } from 'vitest';
+import { render } from 'ink-testing-library';
+import { RenderInline, getPlainTextLength } from './InlineMarkdownRenderer.js';
+import stringWidth from 'string-width';
+
+const stripAnsi = (str: string) =>
```
Comment (jacob314) [inline packages/cli/src/ui/utils/InlineMarkdownRenderer.test.tsx]: why do we have this method here? There are existing libraries for this and I would argue we don't want to strip ansi chars at all in these methods as they block being able to detect whether we accidentally rendered ansi control code. Generally I think it is moot as the output from Ink in lastFrame() doesn't contain ansi control codes under normal circumstances.

File: packages/cli/src/ui/utils/InlineMarkdownRenderer.tsx [hunk 3]
```diff
@@ -6,168 +6,282 @@
...
+ */
+interface MarkdownHandler {
+  render: (match: string, key: string) => React.ReactNode;
+  strip: (match: string) => string;
+}
+
+/**
+ * The map for all markdown operations.
+ */
+const MARKDOWN_HANDLERS: Record<keyof MarkdownMatchGroups, MarkdownHandler> = {
```
Comment (jacob314) [inline packages/cli/src/ui/utils/InlineMarkdownRenderer.tsx:142]: there is a lot of repetition here. Please add a helper so each of the consts is not repeated 4 times.

File: packages/cli/src/ui/utils/InlineMarkdownRenderer.tsx [hunk 4]
```diff
@@ -6,168 +6,282 @@
...
+
+/**
+ * Checks if a string is just plain text, with no markdown or URLs.
+ * This allows for a fast exit before attempting to parse.
+ */
+const isPlainText = (text: string): boolean =>
+  !POTENTIAL_MARKDOWN_REGEX.test(text);
+
+// Marker lengths for slicing
 const BOLD_MARKER_LENGTH = 2; // For "**"
```
Comment (jacob314) [inline packages/cli/src/ui/utils/InlineMarkdownRenderer.tsx]: optional nit: not a new issue but consider refactoring so the consts are the actual markers rather than the lengths.

File: packages/cli/src/ui/utils/InlineMarkdownRenderer.tsx [hunk 5]
```diff
@@ -6,168 +6,282 @@
...
+  let lastIndex = 0;
+
+  for (const match of text.matchAll(MARKDOWN_TOKEN_REGEX)) {
+    const matchIndex = match.index!;
+
+    // Add any plain text before this match
+    addPrecedingPlainTextNode(nodes, text, lastIndex, matchIndex);
+
+    // Get the rendered node for the match itself
+    const key = `m-${matchIndex}`;
```
Comment (jacob314) [inline packages/cli/src/ui/utils/InlineMarkdownRenderer.tsx:250]: I think this code is cargo culting adding keys. Some of your new code doesn't add keys which is fine. I would suggest cleaning up all the keys in this file as they likely don't serve any purpose.

File: packages/cli/src/ui/utils/__snapshots__/InlineMarkdownRenderer.test.tsx.snap [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+exports[`RenderInline > renders a markdown mailto link as text 1`] = `
+"Contact us at
+email (mailto:team@example.com)
+."
+`;
+
+exports[`RenderInline > renders bold text correctly 1`] = `
+"This is
+bold
```
Comment (jacob314) [inline packages/cli/src/ui/utils/__snapshots__/InlineMarkdownRenderer.test.tsx.snap]: these test just verify we strip markdown not that we style it correctly? can we go a little futher?

**Review by jacob314 — COMMENTED** [review]
This is definitely an improvement over the previous inline markdown rendering. approved once the fairly minor comments are addressed.

File: packages/cli/src/ui/utils/InlineMarkdownRenderer.test.tsx (hunk 2 above)
Comment (davideast) [reply packages/cli/src/ui/utils/InlineMarkdownRenderer.test.tsx]: I wasn't exactly sure if we wanted to add another library just to handle this but I'll definitely give the lastFrame() a try without any processing.

File: packages/cli/src/ui/utils/InlineMarkdownRenderer.tsx (hunk 4 above)
Comment (davideast) [reply packages/cli/src/ui/utils/InlineMarkdownRenderer.tsx]: Great idea. I like that.

File: packages/cli/src/ui/utils/__snapshots__/InlineMarkdownRenderer.test.tsx.snap (hunk 1 above)
Comment (davideast) [reply packages/cli/src/ui/utils/__snapshots__/InlineMarkdownRenderer.test.tsx.snap]: I think this is a mistake from me updating the snapshot. I'll double check all these.

File: packages/cli/src/ui/utils/InlineMarkdownRenderer.tsx (hunk 5 above)
Comment (davideast) [reply packages/cli/src/ui/utils/InlineMarkdownRenderer.tsx:250]: `match.index` is stable and unique here because it is the character position of the markdown match within the original static string and is regenerated in the text prop changes.

**Comment by bdmorgan** [comment]
Hi @davideast, thank you so much for your contribution to Gemini CLI! We really appreciate the time and effort you've put into this.

We're making some updates to our contribution process to improve how we track and review changes. Please take a moment to review our recent discussion post: [Improving Our Contribution Process & Introducing New Guidelines](${DISCUSSION_URL}).

Key Update: Starting **January 26, 2026**, the Gemini CLI project will require all pull requests to be associated with an existing issue. Any pull requests not linked to an issue by that date will be automatically closed.

Thank you for your understanding and for being a part of our community!
